extension/src/state_aggregate/rollup.rs

```diff
@@ -0,0 +1,407 @@
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<StateAgg<'a>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let mut state = match state {
+                None => return None,
+                Some(state) => state.clone(),
+            };
+            state.merge();
+            assert!(state.values.len() == 1);
+            let agg: Option<OwnedStateAgg> = state.values.drain(..).next().unwrap().into();
+            agg.map(Into::into)
+        })
+    }
+}
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+fn timeline_agg_rollup_final<'a>(
+    state: Internal,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<TimelineAgg<'a>> {
+    timeline_agg_rollup_final_inner(unsafe { state.to_inner() }, fcinfo)
+}
+
+fn timeline_agg_rollup_final_inner<'a>(
+    state: Option<Inner<RollupTransState>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<TimelineAgg<'a>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let mut state = match state {
+                None => return None,
+                Some(state) => state.clone(),
+            };
+            state.merge();
+            assert!(state.values.len() == 1);
+            let agg: Option<OwnedStateAgg> = state.values.drain(..).next().unwrap().into();
+            agg.map(Into::into).map(TimelineAgg::new)
+        })
+    }
+}
+
+#[pg_extern(immutable, parallel_safe, strict, schema = "toolkit_experimental")]
+pub fn state_agg_rollup_serialize(state: Internal) -> bytea {
+    let mut state: Inner<RollupTransState> = unsafe { state.to_inner().unwrap() };
+    state.merge();
+    crate::do_serialize!(state)
+}
+
+#[pg_extern(strict, immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn state_agg_rollup_deserialize(bytes: bytea, _internal: Internal) -> Option<Internal> {
+    state_agg_rollup_deserialize_inner(bytes).internal()
+}
+pub fn state_agg_rollup_deserialize_inner(bytes: bytea) -> Inner<RollupTransState> {
+    let t: RollupTransState = crate::do_deserialize!(bytes, RollupTransState);
+    t.into()
+}
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn state_agg_rollup_combine(
+    state1: Internal,
+    state2: Internal,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal> {
+    unsafe {
+        state_agg_rollup_combine_inner(state1.to_inner(), state2.to_inner(), fcinfo).internal()
+    }
+}
+
+pub fn state_agg_rollup_combine_inner(
+    state1: Option<Inner<RollupTransState>>,
+    state2: Option<Inner<RollupTransState>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<RollupTransState>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || match (state1, state2) {
+            (None, None) => None,
+            (Some(x), None) => Some(x),
+            (None, Some(x)) => Some(x),
+            (Some(mut x), Some(mut y)) => {
+                x.values.append(&mut y.values);
+                Some(x.into())
+            }
```

Comment:
The inputs to combine are pointers to shared memory and shouldn't be modified or returned, as another thread is free to reuse the memory as soon as this call completes.